examples/rustls-mtls/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+    root_store: rustls::RootCertStore,
+    my_cert_chain: Vec<rustls::Certificate>,
+    my_private_key: rustls::PrivateKey,
+}
+
+impl tls::Provider for MtlsProvider {
+    type Server = Server;
+    type Client = Client;
+    type Error = rustls::Error;
+
+    fn start_server(self) -> Result<Self::Server, Self::Error> {
+        let verifier = rustls::server::AllowAnyAuthenticatedClient::new(self.root_store);
+        let mut cfg = ServerConfig::builder()
+            .with_cipher_suites(DEFAULT_CIPHERSUITES)
+            .with_safe_default_kx_groups()
+            .with_protocol_versions(PROTOCOL_VERSIONS)?
+            .with_client_cert_verifier(verifier)
+            .with_single_cert(self.my_cert_chain, self.my_private_key)?;
+
+        cfg.ignore_client_order = true;
+        cfg.max_fragment_size = None;
+        cfg.alpn_protocols = vec![b"h3".to_vec()];
+        Ok(cfg.into())
+    }
+
+    fn start_client(self) -> Result<Self::Client, Self::Error> {
+        let mut cfg = ClientConfig::builder()
+            .with_cipher_suites(DEFAULT_CIPHERSUITES)
+            .with_safe_default_kx_groups()
+            .with_protocol_versions(PROTOCOL_VERSIONS)?
+            .with_root_certificates(self.root_store)
+            .with_single_cert(self.my_cert_chain, self.my_private_key)?;
+
+        cfg.max_fragment_size = None;
+        cfg.alpn_protocols = vec![b"h3".to_vec()];
+        Ok(cfg.into())
+    }
+}
+
+impl MtlsProvider {
+    pub fn new<A: AsRef<Path>, B: AsRef<Path>, C: AsRef<Path>>(
+        ca_cert_pem: A,
+        my_cert_pem: B,
+        my_key_pem: C,
+    ) -> Result<Self, Error> {
+        let root_store = into_root_store(ca_cert_pem.as_ref())?;
+        let cert_chain = into_certificate(my_cert_pem.as_ref())?;
+        let private_key = into_private_key(my_key_pem.as_ref())?;
+        Ok(MtlsProvider {
+            root_store,
+            my_cert_chain: cert_chain.into_iter().map(rustls::Certificate).collect(),
+            my_private_key: rustls::PrivateKey(private_key),
+        })
+    }
+}
+
+fn into_certificate(path: &Path) -> Result<Vec<Vec<u8>>, Error> {
+    let mut rdr = BufReader::new(
+        std::fs::File::open(path)
+            .map_err(|e| Error::General(format!("Failed to load file: {}", e)))?,
+    );
+    let certs = rustls_pemfile::certs(&mut rdr)
+        .map(|certs| certs.into_iter().collect())
+        .map_err(|_| Error::General("Could not read certificate".to_string()))?;
+    Ok(certs)
+}
+
+fn into_root_store(path: &Path) -> Result<RootCertStore, Error> {
+    let ca_certs = into_certificate(path)?;
+    let mut cert_store = RootCertStore::empty();
+    cert_store.add_parsable_certificates(ca_certs.as_slice());
+    Ok(cert_store)
+}
+
+fn into_private_key(path: &Path) -> Result<Vec<u8>, Error> {
+    let mut f = std::fs::File::open(path)
+        .map_err(|e| Error::General(format!("Failed to load file: {}", e)))?;
+    let mut buf = Vec::new();
+    f.read_to_end(&mut buf)
```

Comment:
Since folks are likely to copy paste these example, I think we should make these async tokio operations. Otherwise we are blocking on the read here.

I think that applies for all IO operations.

Reply:
Done.